File: src/main/scala/io/github/noeppi_noeppi/tools/cfupdatechecker/ChangelogProcessor.scala
package io.github.noeppi_noeppi.tools.cfupdatechecker

import org.apache.commons.text.StringEscapeUtils
import org.jsoup.Jsoup

object ChangelogProcessor {

  def process(html: String): String = normalizeText(Jsoup.parse(html).wholeText())
  
  private def normalizeText(str: String): String = {
    val theStr = if (str.startsWith("\"") && str.endsWith("\"")) str.substring(1, str.length - 1) else str
    StringEscapeUtils.unescapeJava(theStr).replace("\r", "").replaceAll("\\n+", "\n").strip()
      .map(chr => if (chr.isWhitespace && chr != '\n') ' ' else chr).replaceAll(" +", " ")
      .replace("\n \n", "\n").replaceAll("\\n+", "\n").strip()
  }
}

File: src/main/java/org/moddingx/cfupdatechecker/ChangelogProcessor.java
package org.moddingx.cfupdatechecker;

import org.apache.commons.text.StringEscapeUtils;
import org.jsoup.Jsoup;

public class ChangelogProcessor {

    public static String process(String html) {
        return normalizeText(Jsoup.parse(html).wholeText());
    }

    private static String normalizeText(String str) {
        String theStr = str.startsWith("\"") && str.endsWith("\"") ? str.substring(1, str.length() - 1) : str;
        theStr = StringEscapeUtils.unescapeJava(theStr).replace("\r", "").replaceAll("\\n+", "\n").strip();
        StringBuilder builder = new StringBuilder();
        for (char c : theStr.toCharArray()) {
            if (Character.isWhitespace(c) && c != '\n') c = ' ';
            builder.append(c);
        }
        return builder.toString().replaceAll(" +", " ").replace("\n \n", "\n").replaceAll("\\n+", "\n").strip();
    }
}
